### PR TITLE
Enable JSI engine selection in devsettings

### PR DESCRIPTION
**Pull request by tudorms**
We want to begin standing up the JSI based executors and need a way to select the engine during runtime creation. Also move the HermesRuntimeHolder and ChakraJSIRuntimeHolder into ReactWindowsCore to enable use from both win32 and windows.

Nothing changes in the default case with this PR (win32 will continue to use the legacy ChakraExecutor), this is just to support future JSI Runtime implementations.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/2944)

### CODE REVIEW

File: vnext/Shared/OInstance.cpp [hunk 1]
```diff
@@ -430,6 +435,23 @@ InstanceImpl::InstanceImpl(
     if (m_devSettings->jsiRuntimeHolder) {
```
Comment (Khalef1) [inline vnext/Shared/OInstance.cpp]: >(m_devSettings->jsiRuntimeHolder [](start = 7, length = 32)

Should we add an assert to check that this is set only when m_devSettings->jsiEngineOverride == JSIEngineOverride::Default?

File: vnext/Shared/OInstance.cpp [hunk 2]
```diff
@@ -430,6 +435,23 @@ InstanceImpl::InstanceImpl(
...
       jsef = std::make_shared<OJSIExecutorFactory>(
           m_devSettings->jsiRuntimeHolder, m_devSettings->loggingCallback);
+    } else if (m_devSettings->jsiEngineOverride != JSIEngineOverride::Default) {
+      switch (m_devSettings->jsiEngineOverride) {
```
Comment (Khalef1) [inline vnext/Shared/OInstance.cpp]: >s [](start = 6, length = 1)

one more indentation level?

Comment (tudorms) [reply vnext/Shared/OInstance.cpp]: Unfortunately, for some reason we enforce the formatting with clang-format and actually fail the PRs for that reason.

---
In reply to: [314904875](https://github.com/microsoft/react-native-windows/pull/2944#discussion_r314904875) [](ancestors = 314904875)

File: change/react-native-windows-2019-08-16-14-49-31-jsi-engine_override.json [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
+  "type": "none",
+  "comment": "Add jsiEngineOverride option to devsettings",
+  "packageName": "react-native-windows",
+  "email": "tudorm@microsoft.com",
+  "commit": "086aba4eb8a1e1ab0b80c212290a541bf63c2ba9",
+  "date": "2019-08-16T21:49:31.769Z"
+}
```
Comment (Khalef1) [inline change/react-native-windows-2019-08-16-14-49-31-jsi-engine_override.json]: In codeflow I see the the runtimeholder files were moved from vnext/ReactWindowsCore, but I don't see them being deleted from vnext/ReactUWP/base/... . Not sure if this is just the way the tooling works.

Comment (tudorms) [reply change/react-native-windows-2019-08-16-14-49-31-jsi-engine_override.json]: On GitHub they do show up as "move + modify", probably that confuses CodeFlow: [https://github.com/microsoft/react-native-windows/pull/2944/files](https://github.com/microsoft/react-native-windows/pull/2944/files)

---
In reply to: [314905668](https://github.com/microsoft/react-native-windows/pull/2944#discussion_r314905668) [](ancestors = 314905668)

File: vnext/Shared/OInstance.cpp (hunk 2 above)
Comment (NikoAri) [reply vnext/Shared/OInstance.cpp]: There is a macro to turn clang format off for certain sections. We should only enforce it when it makes sense.

---
In reply to: [314905445](https://github.com/microsoft/react-native-windows/pull/2944#discussion_r314905445) [](ancestors = 314905445,314904875)

File: vnext/Shared/OInstance.cpp [hunk 4]
```diff
@@ -428,6 +433,26 @@ InstanceImpl::InstanceImpl(
...
+        case JSIEngineOverride::Hermes:
+#if defined(USE_HERMES)
+          m_devSettings->jsiRuntimeHolder =
+              std::make_shared<HermesRuntimeHolder>();
+          break;
+#else
+          assert(false); // Hermes is not available in this build, fallthrough
+#endif
+        case JSIEngineOverride::Chakra:
+        default: // TODO: Add other engines once supported
```
Comment (NikoAri) [inline vnext/Shared/OInstance.cpp]: >default:  [](start = 8, length = 9)

Shouldn't we crash or have some error for unsupported engines?

File: vnext/ReactWindowsCore/HermesRuntimeHolder.h [hunk 5]
```diff
@@ -1,10 +1,10 @@
 #pragma once
-
 #include <jsi/RuntimeHolder.h>
 #include <jsi/jsi.h>
+#include <thread>
 
+namespace facebook {
```
Comment (NikoAri) [inline vnext/ReactWindowsCore/HermesRuntimeHolder.h]: >{ [](start = 19, length = 1)

nit can we start switching to new "::" notation for namespaces?

**Pull request merged**